### <p style="text-align: right;"> &#9989; **Group 1 (GoGreen)** </p>
#### <p style="text-align: right;"> &#9989; Beth, Zhongjie, McKenna, Erik</p>

# Module 1



## How do SNPs relate to gene transcription?

Downsampled data to toy around. 

Jupyter by default converts `tabs` into `spaces` which is good.

First, import the usual libraries
- `math`: basic math operations
- `os`: enable file manipulation with the OS
- `sys`: enable interaction with commandline
- `glob`: more variable manipulation
- `matplotlib.pyplot`: default plotter (I personally like ggplot waaaaay better. E)
    - `inline`: so that plots are shown in the notebook
- `seaborn`: nicer plots
- `numpy`: all number cruching done here
- `pandas`: data wrangling

In [2]:
import math  
import os   
import glob 
import matplotlib.pyplot as plt 
%matplotlib inline
import seaborn as sns
import numpy as np
import pandas as pd


In [9]:
def dot_to_int(char_in, substitute='.', value=0.0):
    """
    Substitute a string for a given value if such string is of interest
    
    Parameters
    ----------
    char_in : string
         possible char to be substituted
    
    substitute : string
         if the given input matches this object, substitute
    
    value : int
         if applicable, substitute the input with this value
    
    Returns
    -------
    value : if input matches substitute, else returns unchanged input
    """
    if(char_in == substitute):
        return value
    else:
        return char_in
    
def numMissing(row, start=10, missing='.'):
    foo = row[start:] == missing
    return(foo.sum())


def numData(row, data_cols=['Total Calls', 'Missing Data']):
    return row[data_cols].sum()  

Define some functions. Let's see for how long can we keep tidy commented code.

We can print whatever text was written between the `"""` above.

In [6]:
# Set the paths to the raw data
src = '/home/ejam/documents/css893/GoGreen/sample_data/'
dst = '/home/ejam/documents/css893/output/'
file_snp = 'snp_matrix_downsample.csv'
file_gen = 'genes_w_features_downsample.csv'

In [4]:
ATCG = ['Percent A', 'Percent T', 'Percent C', 'Percent G']
#converters = dict((A,dot_to_int) for A in ATCG)

In [7]:
# BEWARE : Line below consumes 24Gb of RAM at some point. 
# snp_matrix = pd.read_table(src + file_snp, dtype={'chromosome': str}, converters=converters)

snp_matrix = pd.read_csv(src + file_snp)

In [8]:
snp_matrix

chromosome  position reference_allele         Allele Counts  \
0     B73V4_ctg10    292327                T  A(0)T(744)C(0)G(105)   
1     B73V4_ctg10    292384                G    A(2)T(0)C(0)G(906)   
2     B73V4_ctg10    292386                G  A(107)T(0)C(0)G(794)   
3     B73V4_ctg10    292402                A    A(912)T(1)C(0)G(0)   
4     B73V4_ctg10    292404                G    A(1)T(0)C(0)G(913)   
..            ...       ...              ...                   ...   
516  B73V4_ctg107     62456                G    A(0)T(0)C(9)G(211)   
517  B73V4_ctg107     62498                G    A(0)T(0)C(1)G(373)   
518  B73V4_ctg107     62530                T  A(0)T(197)C(0)G(163)   
519  B73V4_ctg107     62588                G   A(0)T(0)C(86)G(180)   
520  B73V4_ctg107     63096                G   A(0)T(0)C(11)G(197)   

     Missing Data  Total Calls  Percent A  Percent T  Percent C  Percent G  \
0             110          849   0.000000  87.632509   0.000000  12.367491   
1              51          908   0.220264   0.000000   0.000000  99.779736   
2              58          901  11.875694   0.000000   0.000000  88.124306   
3              46          913  99.890471   0.109529   0.000000   0.000000   
4              45          914   0.109409   0.000000   0.000000  99.890591   
..            ...          ...        ...        ...        ...        ...   
516           739          220   0.000000   0.000000   4.090909  95.909091   
517           585          374   0.000000   0.000000   0.267380  99.732620   
518           599          360   0.000000  54.722222   0.000000  45.277778   
519           693          266   0.000000   0.000000  32.330827  67.669173   
520           751          208   0.000000   0.000000   5.288462  94.711538   

     ... NC328 NC326 PHV53 DKIBC2 A641 WIL900 Va22 E8501 PHP85 Oh43  
0    ...     T     T     .      T    T      .    G     T     T    G  
1    ...     G     G     G      G    G      G    G     G     G    G  
2    ...     G     G     G      G    G      G    A     G     G    A  
3    ...     A     A     A      A    A      A    A     A     A    A  
4    ...     G     G     G      G    G      G    G     G     G    G  
..   ...   ...   ...   ...    ...  ...    ...  ...   ...   ...  ...  
516  ...     G     G     .      .    .      .    .     .     .    G  
517  ...     G     G     .      .    .      .    .     G     .    G  
518  ...     T     T     .      .    .      .    .     .     .    T  
519  ...     G     G     C      .    .      .    .     .     .    G  
520  ...     G     G     .      .    .      .    .     .     .    G  

[521 rows x 952 columns]

First 10 columns are metadata. The rest correspond to cultivars.

The column names for cultivars is found in `B73_plus_RTAs_snp_matrix_imputed/widiv_942g_979873SNPs_imputed_filteredGenos_withRTA_AGPv4.hmp.txt`

In [10]:
missing_data = snp_matrix.apply(numMissing, axis=1).values
discrepancy_missing = snp_matrix['Missing Data'].values - missing_data
print('A total of', len(discrepancy_missing != 0), 'SNPs have different Missing_Data as reported.')
print('Largest discrepancy: ', discrepancy_missing.max(), '[ at row', discrepancy_missing.argmax(),']' )
total_data = snp_matrix.apply(numData, axis=1).values
print('\n\nMean total:\t', total_data.mean())
print('Var total:\t', total_data.var())

A total of 521 SNPs have different Missing_Data as reported.
Largest discrepancy:  17 [ at row 232 ]


Mean total:	 959.0
Var total:	 0.0


The given data assumes **959** total calls (cultivars?), but we only have **942** available cultivars. Thus the `Missing Data` column has **wrong** values.

Now let's explore the expression data!!

In [11]:
genes_w_features = pd.read_csv(src+file_gen)
genes_w_features

gene    chromosome feature_type  position_left  position_right  \
0   Zm00001d000060   B73V4_ctg10         gene         292299          293833   
1   Zm00001d000058   B73V4_ctg10         gene         155738          156620   
2   Zm00001d000072   B73V4_ctg10         gene         917443          925264   
3   Zm00001d000065   B73V4_ctg10         gene         633829          634595   
4   Zm00001d000062   B73V4_ctg10         gene         518226          519667   
5   Zm00001d000071   B73V4_ctg10         gene         856577          871342   
6   Zm00001d000055   B73V4_ctg10         gene          94252           97609   
7   Zm00001d000067   B73V4_ctg10         gene         676059          676732   
8   Zm00001d000064   B73V4_ctg10         gene         605862          608484   
9   Zm00001d000070   B73V4_ctg10         gene         839039          842527   
10  Zm00001d000061   B73V4_ctg10         gene         480282          481819   
11  Zm00001d000066   B73V4_ctg10         gene         655390          657780   
12  Zm00001d000286  B73V4_ctg104    tRNA_gene          42946           43019   
13  Zm00001d000288  B73V4_ctg104         gene          67197           68824   
14  Zm00001d000287  B73V4_ctg104    tRNA_gene          43680           43753   
15  Zm00001d000424  B73V4_ctg105         gene          38569           40097   
16  Zm00001d000301  B73V4_ctg107         gene          28849           31668   
17  Zm00001d000307  B73V4_ctg107         gene          61609           62569   
18  Zm00001d000306  B73V4_ctg107         gene          58889           59216   
19  Zm00001d000304  B73V4_ctg107         gene          45538           50405   
20  Zm00001d000305  B73V4_ctg107         gene          50633           54140   
21  Zm00001d000300  B73V4_ctg107         gene          25721           27077   
22  Zm00001d000302  B73V4_ctg107         gene          36638           40287   
23  Zm00001d000303  B73V4_ctg107         gene          44571           45477   

         LH128      DKMBZA       CQ806      DKF274      Ill.Hy  ...  \
0    28.756000   51.215800   29.783800   22.691200   50.314600  ...   
1     0.000000    0.000000    0.000000    0.000000    0.000000  ...   
2    22.122900   12.008100    3.541750   13.379200    4.843670  ...   
3     0.000000    0.000000    0.570467    0.117619    0.970843  ...   
4     0.000000    0.000000    0.000000    0.000000    0.275438  ...   
5    17.306500    9.370400   12.472600   10.937900    7.867670  ...   
6     0.709536    1.266330    1.322040    0.873182    1.743230  ...   
7     0.000000    0.076162    0.000000    0.163776    0.000000  ...   
8     0.000000   10.168300    0.000000   10.076400    0.000000  ...   
9     4.706050   10.517500   19.070200    6.401030   13.389400  ...   
10    0.349312    0.365797    0.151518    0.219695    0.293829  ...   
11    0.000000    0.098150    0.018979    0.000000    0.000000  ...   
12    0.000000    0.000000    0.000000    0.000000    0.000000  ...   
13    2.988290    2.856140    3.375930    4.082270    3.684240  ...   
14    0.000000    0.000000    0.000000    0.000000    0.000000  ...   
15   18.056200    6.222040   11.157100   23.224800    6.805910  ...   
16    1.677040    1.153180    1.597700    1.371930    0.729046  ...   
17    9.860550    0.028279    5.977300    2.077900    8.896710  ...   
18    0.000000    0.000000   12.048100   14.580700    1.675010  ...   
19    0.000000    0.000000    0.007159    0.000000    0.108632  ...   
20   13.332900   12.656300   12.529900   12.155100   11.140000  ...   
21  107.423000  164.469000  134.920000  163.466000  110.149000  ...   
22   62.563400   67.804300   56.838600   56.477000   70.602800  ...   
23    0.000000    0.000000    0.000000    0.000000    0.000000  ...   

        NC328      NC326      PHV53     DKIBC2       A641    WIL900  \
0   58.436000  54.060700  18.047000   87.40760  43.509200  16.43820   
1    0.000000   0.000000   0.000000    0.00000   0.000000   0.00000   
2    9.396230   9.950220

We have 44259 rows × 947 columns. The set of chromosomes in both data files **are different!!!** **Neither** is a subset of the other. Observe that `'B73V4_ctg100'` shows up only in `genes_w_features`. Both data sets contain 942 cultivars.